lib/cloud/clients.go

```diff
@@ -231,143 +233,145 @@ type azureClients struct {
 	azureRunCommandClients azure.ClientMap[azure.RunCommandClient]
 }
 
-// GetAWSSession returns AWS session for the specified region.
-func (c *cloudClients) GetAWSSession(region string) (*awssession.Session, error) {
-	c.mtx.RLock()
-	if session, ok := c.awsSessions[region]; ok {
-		c.mtx.RUnlock()
-		return session, nil
-	}
-	c.mtx.RUnlock()
-	return c.initAWSSession(region)
-}
-
-// GetAWSSessionForRole returns AWS session for the specified role ARN.
-func (c *cloudClients) GetAWSSessionForRole(ctx context.Context, region, roleARN string) (*awssession.Session, error) {
-	defaultSession, err := c.GetAWSSession(region)
+// GetAWSSession retrieves an AWS session for the given region,
+// filters out roles that have an empty role ARN,
+// assumes each non-empty role ARN in order, and returns an AWS session
+// for the chain of assumed roles.
+//
+// If there are no roles to assume, this function returns the default session
+// for the given region.
+//
+// If multiple AssumeRole arguments are passed, then the roles are assumed in order,
+// i.e. [role1 role2] => assume role1 => assume role2.
+//
+// There are currently two reasons that roles may be assumed:
+//  1. The Teleport database AWS metadata or discovery matcher is configured
+//     to assume a role.
+//  2. Specific protocols (DynamoDB, Redshift Serverless) use AWS assumed roles
+//     for RBAC, where db user is the name of an AWS role to assume.
```

Comment:
yeah we do cache them. I think I need to refactor this, because currently there is no way to distinguish between configured role and --db-user role from cloud clients perspective, and I think we do want to cache the configure role indefinitely

Reply:
Refactoring this to avoid caching db-user roles. I've still got tests to fix, but the implementation will no longer cache them.

Reply:
With cloud clients interface now, it's slightly more tedious to get a session for cassandra/redshift serverless, but overall simplifies things a lot. I don't think we will ever support assuming more than 2 roles, and there's no need to have cloud clients handle an arbitrary length chain of roles. I also don't think it's even worth having the "cache key builder" anymore, and without that I don't need the dependency on lib/services in lib/cloud anymore. Let me know what you think of this interface vs the old one.